tests/specs/lemmas.k

```diff
@@ -330,6 +349,23 @@ module LEMMAS-HASKELL [symbolic, kore]
     rule BA:ByteArray   ==K #buf(32, DATA) => #buf(32, DATA) ==K              BA                                       [simplification, concrete(BA)]
     rule #buf(32, DATA) ==K BA:ByteArray   =>          DATA  ==Int #asInteger(BA) requires #sizeByteArray(BA) <=Int 32 [simplification, concrete(BA)]
 
+    rule #range( M [ N1 := BUF1:ByteArray ] [ N2 := BUF2:ByteArray ] [ N3 := BUF3:ByteArray ], L, K ) =>
+         #range( M                          [ N2 := BUF2           ] [ N3 := BUF3           ], L, K )
+      requires N1 +Int #sizeByteArray(BUF1) <=Int L [simplification]
+
+    rule #range ( M [ N := BUF:ByteArray ], N, K ) => #padToWidth(K, BUF)
+      requires ( #sizeByteArray( BUF ) <=Int K ) andBool ( M [ N .. K ] ==K #padToWidth( K, .ByteArray ) ) [simplification]
+
+    rule #asWord( #padToWidth ( _, BUF ) ) => #asWord( BUF ) [simplification]
+
```

Comment:
Is this true in general? I believe it's only true if `_ <=Int 32`?

Reply:
I'm not sure I understand why it would need that condition. What I was thinking was `#asWord` chops to 32 bytes either way before converting to an Int. If the input is padded with 0s to more than 32 bytes, wouldn't they get `#chop`ed off anyway? And if BUF is already longer than `_`, `#padToWidth` can only add padding, not chop anything.

Reply:
Hmmm, the implementations of `#asWord` that we have are (all from `evm-types.md`):

Java backend:

```
    syntax Int ::= #asWord ( ByteArray ) [function, functional, smtlib(asWord)]
 // ---------------------------------------------------------------------------
    rule [#asWord.base-empty]:  #asWord( .WordStack     ) => 0
    rule [#asWord.base-single]: #asWord( W : .WordStack ) => W
    rule [#asWord.recursive]:   #asWord( W0 : W1 : WS   ) => #asWord(((W0 *Word 256) +Word W1) : WS)
```

And Haskell/LLVM backends:

```
    syntax Int ::= #asWord ( ByteArray ) [function, functional, smtlib(asWord)]
 // ---------------------------------------------------------------------------
    rule #asWord(WS) => chop(Bytes2Int(WS, BE, Unsigned)) [concrete]
```

So it does appear, in both cases, that we end up chopping the resulting integer. The only remaining question is "does the endianess matter here?" We're using Big-Endian encoding for the ByteArray, which indeed seems to do as we want.

`#padToWidth` pads on the left, which will also once again do what we want.

And inspecting `#padToWidth` (LLVM/Haskell implementation):

```
    syntax ByteArray ::= #padToWidth      ( Int , ByteArray ) [function, functional]
 // --------------------------------------------------------------------------------
    rule                        #padToWidth(N, BS)      =>               BS        requires notBool (N >=Int 0)
    rule [padToWidthNonEmpty] : #padToWidth(N, BS)      =>  padLeftBytes(BS, N, 0) requires          N >=Int 0
```

Then inspecting `padLeftBytes` (https://kframework.org/k-distribution/include/kframework/builtin/domains/#bytes-padding), it appears that if the `N` supplied is too small, it will leave the input unchanged.

So I think you're correct! Thank you.